### ISSUE TITLE
Name of default factory method does matter

### ISSUE THREAD

**Issue by Perlence** [python-attrs/attrs#264]
Name of default factory method does matter. It shouldn't be the same as the name of attribute:

```python
>>> @attr.s
... class C(object):
...     x = attr.ib(default=1)
...     y = attr.ib()
...     @y.default
...     def y(self):
...         return self.x + 1
>>> C()
C(x=1)
>>> C().y
<bound method C.y of C(x=1)>
```

**Comment by hynek** [comment]
Hm well yeah, that’s how Python works – if you overwrite an attribute the old one is gone.  What do you suggest us to do ?

**Comment by RonnyPfannschmidt** [comment]
attrs could issue a warning if a decorator based replacement is not named as the function because it indicates an error in usage

it should be noted that such a warning may be considered a problem for exotic declaration

**Comment by Perlence** [comment]
@hynek Maybe a note can be added in the docs 😃
```python
@y.default
def name_does_not_matter_as_long_as_it_is_not_y(self):
    ...
```

**Comment by RonnyPfannschmidt** [comment]
@hynek how about mimicing properties, which return a copy with added attributes

**Comment by Tinche** [comment]
We could also return nothing from the decorator? Although that might be confusing.

**Comment by hynek** [comment]
fixed in #370
